crates/bevy_ecs/src/world/unsafe_world_cell.rs

```diff
@@ -70,75 +72,203 @@ use std::any::TypeId;
 /// }
 /// ```
 #[derive(Copy, Clone)]
-pub struct UnsafeWorldCell<'w>(&'w World);
+pub struct UnsafeWorldCell<'w>(*mut World, PhantomData<(&'w World, &'w UnsafeCell<World>)>);
+
+// SAFETY: `&World` and `&mut World` are both `Send`
+unsafe impl Send for UnsafeWorldCell<'_> {}
+// SAFETY: `&World` and `&mut World` are both `Sync`
+unsafe impl Sync for UnsafeWorldCell<'_> {}
 
 impl<'w> UnsafeWorldCell<'w> {
-    pub(crate) fn new(world: &'w World) -> Self {
-        UnsafeWorldCell(world)
+    pub(crate) fn new_readonly(world: &'w World) -> Self {
+        UnsafeWorldCell(world as *const World as *mut World, PhantomData)
     }
 
-    /// Gets a reference to the [`&World`](crate::world::World) this [`UnsafeWorldCell`] belongs to.
-    /// This can be used to call methods like [`World::contains_resource`] which aren't exposed and but don't perform any accesses.
+    pub(crate) fn new_mutable(world: &'w mut World) -> Self {
+        // SAFETY: raw pointer is derived from an `&mut World` and `Unsafecell<World>` has
+        //         the same layout as `World`.
+        unsafe {
+            let world: &'w UnsafeCell<World> = &*(world as *mut World as *mut UnsafeCell<World>);
+            Self(world as *const UnsafeCell<World> as *mut World, PhantomData)
+        }
+    }
+
+    /// This `&mut World` counts as a mutable borrow of every resource and component for the purposes
+    /// of safety invariants that talk about borrows on component/resource data not existing. This is
+    /// true even for methods that do not explicitly call out `&mut World` as problematic.
     ///
-    /// **Note**: You *must not* hand out a `&World` reference to arbitrary safe code when the [`UnsafeWorldCell`] is currently
-    /// being used for mutable accesses.
+    /// # Safety
+    /// - there must be no borrows on world
+    pub unsafe fn world_mut(self) -> &'w mut World {
+        // SAFETY:
+        // - caller ensures the created `&mut World` is the only borrow of world
+        unsafe { &mut *self.0 }
+    }
```

Comment:
This is UB if the `UnsafeWorldCell` was created through `new_readonly`, even if that shared reference was the only reference of the `World`.